ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,363 @@
+
+        Variables Needed:
+            dataset_name (str): the name of the dataset (provided by 'create' input)
+            created_by (str): name of who made the dataset (provided by 'create' input)
+            comments (str): comments or notes supplied by the user regarding the
+                dataset produced by this tool ((provided by 'create' input))
+            training_type (str): the training type selected by the user (provided by 'create' input)
+            image_ids (list): a list of image IDs that ended up in the final
+                dataset (either dev or test) (provided by 'create' input)
+            filters (dict): a dictionary representing filter metadata (provided by 'interactive_filter')
+            dataset_path (Path): where metadata will be written (provided by 'create' input)
+        """
+        dataset_path = self.dataset_path / self.dataset_name
+        metadata_filepath = dataset_path / 'metadata.json'
+
+        metadata = {}
+        metadata["name"] = self.dataset_name
+        metadata["date_created"] = datetime.utcnow().isoformat() + "Z"
+        metadata["created_by"] = self.created_by
+        metadata["comments"] = self.comments
+        metadata["training_type"] = self.plugin_name
+        metadata["image_ids"] = [(image_id[0].name, image_id[1]) for image_id in self.image_ids]
+        metadata["filters"] = self.filter_metadata
+        with open(metadata_filepath, 'w') as outfile:
+            json.dump(metadata, outfile) 
+
+    def write_dataset(self, obj_list: list):
+        """Method is parent function for writing out complete dataset. Method first
+            creates 'test' and 'dev' subsets. The 'test' subset gets all related files
+            to it copied into a test folder. The 'dev' subset calls 'write_out_complete_set'
+            in the 'splits/complete' directory.
+
+            If overloaded, there are no expectations, but note that the variables 'kfolds'
+            and 'test_percent' are provided for use.
+
+        Args:
+            obj_list (list): list of objects to be written in dataset
+
+        Variables Needed:
+            dataset_path (Path): where dataset will be written (provided by 'create' input)
+            dataset_name (str): the name of the dataset (provided by 'create' input)
+            temp_dir (Path): where the raw data files are (provided by 'create' input)
+            associated_files (dict): decides what files are to be copied for the test set
+                (provided by 'create' input)
+        """
+        dataset_path = self.dataset_path / self.dataset_name
+        print(dataset_path)
+
+        test_subset, dev_subset = split_data(obj_list, test_percent=self.test_percent)
+
+        # Test subset
+        test_path = dataset_path / 'test'
+        os.mkdir(test_path)
+        test_image_ids = [(self.temp_dir, obj['image_id']) for obj in test_subset]
+        copy_data_locally(test_image_ids, test_path, self.associated_files)
```

Comment:
I think this default implementation of `write_dataset` could use a bit more work. It's so close to being completely agnostic to the exact way that the plugin constructs its data, but these two lines force the plugin to make every object a dictionary with an 'image_id' key. Since the only reason for this is for the test set, perhaps the creation of the test set should be split out into its own overridable method (just like `write_data`). The default implementation could rely on the objects having an `image_id` attribute, but the plugin could override it without having to override `write_dataset`.